### PR TITLE
Allow older receipts to be viewed at their direct URL

### PR DESCRIPTION
**Pull request by marcinbot**
Work related to p7bbVw-2nJ-p2

~~Requires D10212-code to be shipped before this.~~ The backend changes have been deployed

To test:
* attempt to load an old receipt (one that is not included in the first 200 that are fetched on the history page) by navigating directly to its URL (`/me/purchases/billing/{receiptId}`)
* the page should load
* the page should still redirect if the receipt doesn't exist

### CODE REVIEW

File: client/state/billing-transactions/actions.js [hunk 1]
```diff
@@ -44,6 +47,31 @@ export const requestBillingTransactions = () => {
...
+		.then( receipt => {
+			dispatch( {
+				type: BILLING_RECEIPT_REQUEST_SUCCESS,
+				receiptId,
+				receipt,
+			} );
+		} )
+		.catch( () => {
+			dispatch( {
+				type: BILLING_RECEIPT_REQUEST_FAILURE,
```
Comment (jeffstieler) [inline client/state/billing-transactions/actions.js]: With this, we're treating request errors (like timeouts) the same as permission errors (trying to view someone else's receipt). What do you think about differentiating between the two?

If the receipt request fails due to timeout - should we show a notice with a retry action?

Comment (marcinbot) [reply client/state/billing-transactions/actions.js]: Added notifications:
When a non-existent or someone else's receipt is requested:
screen shot 2018-03-16 at 14 04 14

When connection fails:
screen shot 2018-03-16 at 14 18 15

File: client/me/billing-history/receipt.jsx [hunk 2]
```diff
@@ -58,9 +62,26 @@ class BillingReceipt extends React.Component {
 			transaction,
 			transactionFetchError,
 			transactionId,
+			translate,
 		} = this.props;
 
 		if ( transactionFetchError ) {
+			if ( 'invalid_receipt' === transactionFetchError.error ) {
+				this.props.errorNotice(
+					translate( 'Could not find Receipt #%s', { args: transactionId } ),
```
Comment (jeffstieler) [inline client/me/billing-history/receipt.jsx]: Based on a [search for other error messages](https://github.com/Automattic/wp-calypso/search?q=errorNotice(&type=&utf8=%E2%9C%93), I think this message could be a little more conversational.

How about:
> Sorry, we couldn't find receipt #123.

File: client/me/billing-history/receipt.jsx [hunk 3]
```diff
@@ -58,9 +62,26 @@ class BillingReceipt extends React.Component {
...
+					{
+						displayOnNextPage: true,
+					}
+				);
+			} else {
+				this.props.clearBillingTransactionError( transactionId );
+				this.props.errorNotice( translate( 'There was a problem when fetching the receipt' ), {
```
Comment (jeffstieler) [inline client/me/billing-history/receipt.jsx]: Looking at [other error messages](https://github.com/Automattic/wp-calypso/search?p=3&q=errorNotice(&type=&utf8=%E2%9C%93), I'd like to propose:

> Sorry, we had a problem loading that receipt.

Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Changed both in c675dc4

File: client/state/billing-transactions/reducer.js [hunk 4]
```diff
@@ -46,6 +49,48 @@ export const requesting = createReducer( false, {
...
+		[ BILLING_RECEIPT_REQUEST ]: ( state, { receiptId } ) => ( {
+			...state,
+			requesting: {
+				...state.requesting,
+				[ receiptId ]: true,
+			},
+		} ),
+		[ BILLING_RECEIPT_REQUEST_FAILURE ]: ( state, { receiptId, error } ) => ( {
+			...state,
+			errors: {
```
Comment (tyxla) [inline client/state/billing-transactions/reducer.js]: Seems like using the notices middleware could prevent us from storing the errors in the Redux state completely.

File: client/state/billing-transactions/reducer.js [hunk 5]
```diff
@@ -46,6 +49,48 @@ export const requesting = createReducer( false, {
...
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export const individualTransactions = createReducer(
+	{},
+	{
+		[ BILLING_RECEIPT_REQUEST ]: ( state, { receiptId } ) => ( {
+			...state,
+			requesting: {
```
Comment (tyxla) [inline client/state/billing-transactions/reducer.js]: In the original reducer for all billing transactions we have separate reducers for tracking request state and storing transaction data. Seems like we use the same reducer for that here. 

My personal opinion is that if we keep reducers separate, it would allow us to use `keyedReducer` and keep the code with more separate concerns, and arguably more consistent and easier to test.

Would love to hear your thoughts about this.

File: client/state/billing-transactions/test/actions.js [hunk 6]
```diff
@@ -162,4 +170,71 @@ describe( 'actions', () => {
 			} );
 		} );
 	} );
+
+	describe( '#requestBillingTransaction', () => {
```
Comment (tyxla) [inline client/state/billing-transactions/test/actions.js]: Would be great if we could use jest for these tests. Not a requirement though, as previous tests seem to haven't been migrated yet.

File: client/state/billing-transactions/test/reducer.js [hunk 7]
```diff
@@ -235,4 +244,41 @@ describe( 'reducer', () => {
 			expect( state ).to.eql( {} );
 		} );
 	} );
+
+	describe( '#individualTransactions', () => {
```
Comment (tyxla) [inline client/state/billing-transactions/test/reducer.js]: Reducer tests could use some tests for the cases when we already have data for a certain `receiptId` and we want to test whether it overrides it properly - proposed tests only test against an empty state tree.

File: client/state/billing-transactions/test/reducer.js [hunk 8]
```diff
@@ -235,4 +244,41 @@ describe( 'reducer', () => {
...
+		} );
+
+		test( 'BILLING_RECEIPT_REQUEST_FAILURE', () => {
+			const state = individualTransactions( currentState, {
+				type: BILLING_RECEIPT_REQUEST_FAILURE,
+				receiptId,
+				error: true,
+			} );
+
+			expect( state.requesting[ receiptId ] ).to.eql( false );
```
Comment (tyxla) [inline client/state/billing-transactions/test/reducer.js]: Here you can see a good example of why it would be better if we keep the `requesting` and `items` for the individual transactions in separate reducers.

File: client/state/selectors/get-past-billing-transaction-error.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * Returns a past billing transaction.
+ * Returns null if the billing transactions have not been fetched yet, or there is no transaction with that ID.
+ *
+ * @param  {Object}  state   Global state tree
+ * @param  {String}  id      ID of the transaction
+ * @return {?Object}         The transaction object
+ */
+const getPastBillingTransactionError = ( state, id ) => {
+	return get( state, [ 'billingTransactions', 'individualTransactions', 'errors', id ], false );
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: The documentation implies that if there is no error for the transaction ID we return a `null` but here we return a boolean `false`. Might make more sense to alter the lodash `get` default value to `null` instead, and update tests to comply with that.

File: client/state/selectors/get-past-billing-transaction-error.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * External dependencies
+ */
+
+import { get } from 'lodash';
+
+/**
+ * Returns a past billing transaction.
+ * Returns null if the billing transactions have not been fetched yet, or there is no transaction with that ID.
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: Seems like the description here doesn't match what this selector does.

File: client/state/selectors/get-past-billing-transaction-error.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * @param  {Object}  state   Global state tree
+ * @param  {String}  id      ID of the transaction
+ * @return {?Object}         The transaction object
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: I find this a little confusing, because looking at the tests we only return `bool` values. Looking at the code, it seems we expect error objects. Probably we need to update tests to reflect this.

File: client/state/selectors/get-past-billing-transaction-error.js [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+ */
+
+import { get } from 'lodash';
+
+/**
+ * Returns a past billing transaction.
+ * Returns null if the billing transactions have not been fetched yet, or there is no transaction with that ID.
+ *
+ * @param  {Object}  state   Global state tree
+ * @param  {String}  id      ID of the transaction
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: Is the `ID` a string, or should it be corrected to a `Number` instead?

File: client/state/selectors/get-past-billing-transaction-error.js [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+ * @return {?Object}         The transaction object
+ */
+const getPastBillingTransactionError = ( state, id ) => {
+	return get( state, [ 'billingTransactions', 'individualTransactions', 'errors', id ], false );
+};
+
+export default getPastBillingTransactionError;
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: Seems like we don't need a named export here, a `export default ( ... ) => ...` could be shorter.

File: client/state/selectors/get-past-billing-transaction-error.js (hunk 9 above)
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction-error.js]: Usually when we only have an arrow function only return something in the function block, we could shorten it from `() => { return ... }` to `() => ...`.

File: client/state/selectors/get-past-billing-transaction.js [hunk 14]
```diff
@@ -20,8 +20,17 @@ import getPastBillingTransactions from './get-past-billing-transactions';
...
+const getPastBillingTransaction = createSelector(
+	( state, id ) => {
+		return (
+			find( getPastBillingTransactions( state ), { id } ) ||
+			get( state, [ 'billingTransactions', 'individualTransactions', 'receipts', id ], null )
+		);
+	},
+	state => [
+		getPastBillingTransactions( state ),
+		...keys( get( state, [ 'billingTransactions', 'individualTransactions', 'receipts' ] ) ),
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction.js]: I think this will be calculated every time, regardless of whether the receipt have changed. Can we optimize it?

File: client/state/selectors/get-past-billing-transaction.js [hunk 15]
```diff
@@ -20,8 +20,17 @@ import getPastBillingTransactions from './get-past-billing-transactions';
...
  * @return {?Object}         The transaction object
  */
-const getPastBillingTransaction = createSelector( ( state, id ) => {
-	return find( getPastBillingTransactions( state ), { id } ) || null;
-}, getPastBillingTransactions );
+const getPastBillingTransaction = createSelector(
+	( state, id ) => {
+		return (
+			find( getPastBillingTransactions( state ), { id } ) ||
+			get( state, [ 'billingTransactions', 'individualTransactions', 'receipts', id ], null )
```
Comment (tyxla) [inline client/state/selectors/get-past-billing-transaction.js]: Seems like we could benefit from a more concrete selector for fetching an individual transaction here.

File: client/state/selectors/is-requesting-billing-transaction.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { isRequestingBillingTransactions } from 'state/selectors';
+
+/**
+ * Returns true if we are currently making a request to get the billing transactions.
```
Comment (tyxla) [inline client/state/selectors/is-requesting-billing-transaction.js]: This needs to be updated to correspond to the fact that this selector is responsible for requesting single transactions, and not for requesting all.

File: client/state/selectors/test/get-past-billing-transaction-error.js [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+ * Internal dependencies
+ */
+import { getPastBillingTransactionError } from 'state/selectors';
+
+describe( 'getPastBillingTransactionError()', () => {
+	const state = {
+		billingTransactions: {
+			individualTransactions: {
+				errors: {
+					123: false,
```
Comment (tyxla) [inline client/state/selectors/test/get-past-billing-transaction-error.js]: As mentioned above, it would probably make more sense if we use error objects here, rather than `bool` values. This would make the tests more self-explicable and would document the code a bit better.

File: client/state/selectors/test/get-past-billing-transaction.js [hunk 18]
```diff
@@ -57,4 +58,22 @@ describe( 'getPastBillingTransaction()', () => {
 		);
 		expect( output ).to.be.null;
 	} );
+
+	test( 'should return a billing transaction that has been fetched individually', () => {
```
Comment (tyxla) [inline client/state/selectors/test/get-past-billing-transaction.js]: Do we want to add some more tests for cases where transactions are not being fetched, or are missing?

**Review by tyxla — APPROVED** [review]
Left some feedback, mostly implementational stuff and small details.

I must admit the PR is a little difficult to review because it introduces so much functionality in a single PR. To make review easier (and possibly reduce the chance to miss some edge cases) I'd probably split it to separate PRs as follows:

* Actions, action creators + `UndocumentedMe` new method + their tests
* Reducer(s) + their tests
* Selector(s) + their tests (each selector could be a separate PR too)
* Implementing the new Redux infrastructure into the billing history components

Regardless, this is great work - thanks for working on it, seems really useful 👍 ❤️

**Review by tyxla — COMMENTED** [review]
I inadvertently approved when I actually wanted to comment.

**Comment by tyxla** [comment]
> To make review easier (and possibly reduce the chance to miss some edge cases) I'd probably split it to separate PRs as follows:

Just to mention that this is a noble cause and if you'd like to go that way, I'd be happy to help with reviews all those PRs 😉

**Review by tyxla — CHANGES_REQUESTED** [review]
See above.

File: client/me/billing-history/receipt.jsx [hunk 19]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+				this.props.clearBillingTransactionError( transactionId );
+				this.props.errorNotice( translate( 'Sorry, we had a problem loading that receipt.' ), {
+					displayOnNextPage: true,
+					button: translate( 'Try again' ),
+					href: billingHistoryReceipt( transactionId ),
+				} );
+			}
 
-		if ( ! transaction && totalTransactions !== null ) {
 			page.redirect( billingHistory );
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Yes. It will show the notice

File: client/state/billing-transactions/actions.js [hunk 20]
```diff
@@ -44,6 +45,38 @@ export const requestBillingTransactions = () => {
...
+		.catch( error => {
+			dispatch( {
+				type: BILLING_RECEIPT_REQUEST_FAILURE,
+				receiptId,
+				error,
+			} );
+		} );
+};
+
+export const clearBillingTransactionError = receiptId => ( {
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/state/billing-transactions/actions.js]: This is used to remove the error from the state tree so that when "Try again" is clicked on the notice, it will not cause a redirect.

I'm looking into fixing the redux state, so might still remove this.

[the remark this answers is not in the document]
Comment (tyxla) [reply client/state/billing-transactions/actions.js]: Using the notices middleware would also help you to get rid of it completely.

[the remark this answers is not in the document]
Comment (marcinbot) [reply client/state/billing-transactions/actions.js]: Could you explain how? I can see how I wouldn't need to store the error or error flag in the state if I just want to display notices, but how do I detect a fetch error in order to redirect? And then ignore that error on retry? We don't usually call `page.redirect` from Redux, we do it in the components. Is there maybe a middleware that I'm missing?

File: client/lib/wpcom-undocumented/lib/me.js [hunk 21]
```diff
@@ -47,11 +47,12 @@ UndocumentedMe.prototype.billingHistoryEmailReceipt = function( receiptId, callb
 	return this.wpcom.req.get( args, callback );
 };
 
-UndocumentedMe.prototype.getReceipt = function( receiptId, fn ) {
+UndocumentedMe.prototype.getReceipt = function( receiptId, query, fn ) {
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/lib/wpcom-undocumented/lib/me.js]: Not for how this function is being used. I removed it in 623a48d

File: client/me/billing-history/receipt.jsx [hunk 22]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+			totalTransactions,
+			transaction,
+			transactionFetchError,
+			transactionId,
+			translate,
+		} = this.props;
+
+		if ( transactionFetchError ) {
+			if ( 'invalid_receipt' === transactionFetchError.error ) {
+				this.props.errorNotice(
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Added in 5c0a144

File: client/me/billing-history/receipt.jsx [hunk 23]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+					translate( "Sorry, we couldn't find receipt #%s", { args: transactionId } ),
+					{
+						displayOnNextPage: true,
+					}
+				);
+			} else {
+				this.props.clearBillingTransactionError( transactionId );
+				this.props.errorNotice( translate( 'Sorry, we had a problem loading that receipt.' ), {
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Added in 5c0a144

File: client/me/billing-history/receipt.jsx [hunk 24]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+		if ( transactionFetchError ) {
+			if ( 'invalid_receipt' === transactionFetchError.error ) {
+				this.props.errorNotice(
+					translate( "Sorry, we couldn't find receipt #%s", { args: transactionId } ),
+					{
+						displayOnNextPage: true,
+					}
+				);
+			} else {
+				this.props.clearBillingTransactionError( transactionId );
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: See https://github.com/Automattic/wp-calypso/pull/22442#discussion_r176018848 - not sure how to remove this without using other bad practices.

File: client/me/billing-history/receipt.jsx [hunk 25]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+				} );
+			}
 
-		if ( ! transaction && totalTransactions !== null ) {
 			page.redirect( billingHistory );
+			return;
+		}
+
+		if ( ! transaction && totalTransactions !== null && ! requestingBillingTransaction ) {
+			this.props.requestBillingTransaction( transactionId );
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Added in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-9644488c9470b0449d5a946ca5533155

File: client/me/billing-history/receipt.jsx [hunk 26]
```diff
@@ -285,10 +327,15 @@ export default connect(
 
 		return {
 			transaction: getPastBillingTransaction( state, ownProps.transactionId ),
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: 2cb53e7

File: client/state/action-types.js [hunk 27]
```diff
@@ -75,6 +75,9 @@ export const AUTOMATED_TRANSFER_STATUS_SET = 'AUTOMATED_TRANSFER_STATUS_SET';
 export const BILLING_RECEIPT_EMAIL_SEND = 'BILLING_RECEIPT_EMAIL_SEND';
 export const BILLING_RECEIPT_EMAIL_SEND_FAILURE = 'BILLING_RECEIPT_EMAIL_SEND_FAILURE';
 export const BILLING_RECEIPT_EMAIL_SEND_SUCCESS = 'BILLING_RECEIPT_EMAIL_SEND_SUCCESS';
+export const BILLING_RECEIPT_REQUEST = 'BILLING_RECEIPT_REQUEST';
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/state/action-types.js]: 0701ee7

File: client/me/billing-history/receipt.jsx [hunk 28]
```diff
@@ -45,11 +55,39 @@ class BillingReceipt extends React.Component {
...
+			transactionId,
+			translate,
+		} = this.props;
+
+		if ( transactionFetchError ) {
+			if ( 'invalid_receipt' === transactionFetchError.error ) {
+				this.props.errorNotice(
+					translate( "Sorry, we couldn't find receipt #%s", { args: transactionId } ),
+					{
+						displayOnNextPage: true,
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Added in https://github.com/Automattic/wp-calypso/pull/22442/commits/5c0a1442ba1f746cc601853dfe3643c355c2b2a6#diff-93de61409d00038592a9ad85d7ec7d74R298

File: client/me/billing-history/receipt.jsx [hunk 29]
```diff
@@ -243,7 +285,7 @@ class BillingReceipt extends React.Component {
 				<Card compact className="billing-history__receipt-links">
 					<a
 						href={ transaction.support }
-						className="button is-primary"
+						className="billing-history__support-button button is-primary"
```
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/me/billing-history/receipt.jsx]: Good suggestion! I've implemented it in f3887a2

File: client/state/billing-transactions/reducer.js (hunk 4 above)
Comment (marcinbot) [reply client/state/billing-transactions/reducer.js]: See https://github.com/Automattic/wp-calypso/pull/22442#discussion_r176018848 - still need to store an error in order to redirect. Otherwise I would need to use something like internal component state.

File: client/state/billing-transactions/reducer.js (hunk 5 above)
Comment (marcinbot) [reply client/state/billing-transactions/reducer.js]: Using `keyedReducer` helps and removes the need for extraneous logic. I added it in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-b8ba8cf4f338be62232e1ff4a890823a

File: client/state/billing-transactions/test/reducer.js (hunk 7 above)
Comment (marcinbot) [reply client/state/billing-transactions/test/reducer.js]: Added in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-d71637a2434a4085549a6e10b8d0728a

File: client/state/selectors/get-past-billing-transaction-error.js (hunk 9 above)
Comment (marcinbot) [reply client/state/selectors/get-past-billing-transaction-error.js]: Fixed this selector in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-20d77f0995aab8ead798a5efd3697d4d

File: client/state/selectors/get-past-billing-transaction.js (hunk 14 above)
Comment (marcinbot) [reply client/state/selectors/get-past-billing-transaction.js]: Now with `keyedReducer` I replaced this with a single `get` call: https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-04441a884f49a26d00e1abd2bde67001

File: client/state/selectors/get-past-billing-transaction.js (hunk 15 above)
Comment (marcinbot) [reply client/state/selectors/get-past-billing-transaction.js]: I added it as a private util in the same file: https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-04441a884f49a26d00e1abd2bde67001

With logic of fetching and searching the existing transactions first I would rather not expose this  selector.

File: client/state/selectors/is-requesting-billing-transaction.js (hunk 16 above)
Comment (marcinbot) [reply client/state/selectors/is-requesting-billing-transaction.js]: Updated in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-42cba6160a6ecebf875a62f21ea2f926

File: client/state/selectors/test/get-past-billing-transaction-error.js (hunk 17 above)
Comment (marcinbot) [reply client/state/selectors/test/get-past-billing-transaction-error.js]: The reducer is back to storing `bool`

File: client/state/selectors/test/get-past-billing-transaction.js (hunk 18 above)
Comment (marcinbot) [reply client/state/selectors/test/get-past-billing-transaction.js]: Added in https://github.com/Automattic/wp-calypso/pull/22442/commits/2c04630b746b8513ec105c3a93b44de95c888100#diff-d90cb43cbbf9714706f676acb170ed54

File: client/lib/wpcom-undocumented/lib/me.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (marcinbot) [reply client/lib/wpcom-undocumented/lib/me.js]: Actually, it was being used in https://github.com/Automattic/wp-calypso/blob/master/client/state/receipts/actions.js#L16 but wpcom lib is determining whether the second param is a `query` or `fn` based on its type. Since all calls to this function pass two arguments and it's better to use Promises instead of callbacks, I renamed `query` to `queryOrCallback` in f966722

File: client/state/billing-transactions/actions.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (tyxla) [reply client/state/billing-transactions/actions.js]: I get what you mean - it seems we're handling different errors separately here. There isn't much we can do in that case, unless we migrate to the data layer (which as mentioned is a larger task and is beyond this PR).

So I've reconsidered and think it makes sense to keep this one given the circumstances.

**Comment by tyxla** [comment]
Thanks so much for the thoroughness and patience @marcinbot! I'll have another look as soon as I get a chance.

> I appreciate the idea to split this into multiple PRs, but this is still a rather small change. 

If you ask for my opinion, I strongly disagree here:



I think this is by no means a small change. The impact might arguably be small here, but the change is large, because in this PR we're doing so many things:

* Introducing action creators for individual transactions, together with UndocumentedMe changes
* Introducing reducer for individual transactions
* Updating past transaction selector to work with individual transactions
* Introducing selectors for past billing transaction error and requesting transaction
* Introducing a query component for single querying transactions
* Introducing error notices in the notices middleware
* Updating receipt to work with individual transactions
* Tests for much of the above

Personally, I'd split this to at least 5-6 PRs. But that's me 🤷‍♂️

**Comment by sirreal** [comment]
It looks like this PR is adding some legacy-style tests. We've been using Jest for a while and it would be great to use its built-in [`expect`](https://facebook.github.io/jest/docs/en/expect.html) (it's a global you don't need to import in test files). The test output is considerably better than when using chai `expect`.

It would be great to ensure new tests use [current best practices](https://github.com/Automattic/wp-calypso/blob/master/docs/testing/unit-tests.md).

Here's a PR updating some tests that may help updating: #23482

**Comment by tyxla** [comment]
> Ok, I'll look into splitting it.

Much appreciated @marcinbot! I'd be happy to help with reviews on those PRs, so please ping me 😉

**Comment by tyxla** [comment]
Thanks @marcinbot! But it seems that #23498, #23499, #23500 aren't created off the `master` branch 😕 Also, could I suggest that test instructions are added to each of the PRs? Would be really helpful for other folks to jump in and review and/or test the PRs.

**Comment by ockham** [comment]
In Calypso, we [generally base all branches off `master`](https://github.com/Automattic/wp-calypso/blob/master/docs/git-workflow.md#branch-naming-scheme) -- let's definitely also follow that pattern here. If PRs are modular enough (e.g. actions separate from query components), that should work even if e.g. a query component obviously requires an action to work. We can simply annotate in PR descriptions -- it's what we've done for plenty of other Redux state related PRs before 😉

**Comment by ockham** [comment]
Thanks a lot @marcinbot, appreciate it!
